### DISCUSSION TITLE
The use of + suffix to append new files in plain request

### DISCUSSION THREAD

**Discussion by anonimusprogramus** [pocketbase/pocketbase#6077] [Q&A] [answered]
I'm continuing this [misplaced question](https://github.com/pocketbase/js-sdk/issues/326) (sorry :v:).

I'm using Node v20.10.0 on Linux Mint, with Pocketbase v0.23.5.

My collection structure is just one "upload" collection,

![2024-12-10 19-12-54](https://github.com/user-attachments/assets/be1c993f-a776-4cc8-b0ca-a8d97623916b)

The context I forgot to mention is that I'm making **every upload has its own progressbar**; thus why I'm sending separate request for each file in a loop.

My question would be how to translate
```typescript
await pb.collection('upload').update(uploadId, {
  "files+": new File(["content 3..."], "file3.txt")
});
```

into xhr/fetch PATCH command to ```http://127.0.0.1:8090/api/collections/upload/records/uploadId```

thanks

**Comment by ganigeorgiev** [comment]
It is still not clear what actually is the issue you are experiencing. The provided details are even more confusing because `XMLHttpRequest` as far as I know is a browser only API and I'm not sure how it works in your case with Node.js. Unless you are loading a polyfill? If that's the case then it is possible to be something related to the specific polyfill, I'm not really sure.

Please try to elaborate more clearly how do you send the request and what actually is the issue with your old code, aka. what error is returned if there is any.

There shouldn't be any difference in the submitted FormData payload between the SDK's `fetch` and the `XMLHttpRequest` requests.

**Reply by ganigeorgiev** [reply]
One thing to note is that in your `xhr.open` call you are not specifying the async flag, which means that by default `xhr.send` will return immediately without awaiting for the request to complete, aka. it is possible a following update request(s) for the same record to overwrite a previous one because the server processes the requests concurrently without being aware of the other update requests.

You can try adding an explicit `false` value for the 3rd `xhr.open` argument:

```js
xhr.open(
  "PATCH",
  "http://localhost:8090/api/collections/upload/records/" + uploadId,
  false,
);
```

_(If you are running the above in the browser, note that Firefox doesn't allow synchronous requests on the main thread, see the note in https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/open#async)_

It depends what actually is the desired behavior, but another option if you want to be able to send requests in async manner is to change your collection structure and instead of having a single record with multiple `file` field to have multiple records each with a single `file` field and then the order of the requests processing shouldn't matter.

**Reply by anonimusprogramus** [reply]
Oh OK.

I use React+Vite+client-side Pocketbase SDK. No server, just pure Pocketbase :fire: 
 
And this is how I call the function,
```typescript
async function onSubmit(data: NewUpload) {
  try {
    const { id } = await pb.collection("upload").create({
      name: "Reading part 1",
      description: "Contains 5 pdf",
      user: pb.authStore.record?.id,
    });

    uploadFiles({
      uploadId: id,
      files: data.files,
    });
  } catch (error) {
    console.log({ is: "onSubmit error", error });
  }
}
```
No error made, but only the result is not what's expected.

- Uploading 5 files with SDK, got result 5 files saved (checked in admin dashboard).
- Uploading 5 files with plain request, sometimes get 1 or 2

According to [the doc](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/open), 3rd param `xhr.open` defaulted to async, and for multipart should be true. I tried to set it to false, no different result.

So perhaps since PATCH doesn't append files but replace, it's about racing condition, latest uploaded file win?

Thanks

**Reply by ganigeorgiev** [reply, marked as answer]
>  it's about racing condition, latest uploaded file win?

Yes, most likely. Change your collections structure so that a single `upload` record would have only 1 file or simply submit all files together.

**Reply by anonimusprogramus** [reply]
Now I can rest :smile: 

Thanks a lot
